File: regalia-moshi/src/main/java/com/commit451/regalia/moshi/RealmListJsonAdapterFactory.java
package com.commit451.regalia.moshi;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.Set;

import io.realm.RealmList;
import io.realm.RealmObject;

/**
 * Enables Moshi to parse and serialize {@link io.realm.RealmList}s. Install via your {@link com.squareup.moshi.Moshi.Builder}
 */
public final class RealmListJsonAdapterFactory implements JsonAdapter.Factory {

    @Override
    @Nullable
    public JsonAdapter<?> create(@NonNull Type type, @NonNull Set<? extends Annotation> annotations, @NonNull Moshi moshi) {
        Class<?> rawType = Types.getRawType(type);
        if (!annotations.isEmpty()) return null;
        if (rawType == RealmList.class) {
            return newRealmListAdapter(type, moshi).nullSafe();
        }
        return null;
    }

    private static <T extends RealmObject> JsonAdapter<RealmList<T>> newRealmListAdapter(Type type, Moshi moshi) {
        Type elementType = Types.collectionElementType(type, RealmList.class);
        JsonAdapter<T> elementAdapter = moshi.adapter(elementType);
        return new RealmListAdapter<>(elementAdapter);
    }
}

File: regalia-moshi/src/main/java/com/commit451/regalia/moshi/RealmListJsonAdapterFactory.kt
package com.commit451.regalia.moshi

import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import io.realm.RealmList
import io.realm.RealmObject
import java.lang.reflect.Type

/**
 * Enables Moshi to parse and serialize [io.realm.RealmList]s. Install via your [com.squareup.moshi.Moshi.Builder]
 */
class RealmListJsonAdapterFactory : JsonAdapter.Factory {
    override fun create(type: Type, annotations: Set<Annotation?>, moshi: Moshi): JsonAdapter<*>? {
        val rawType = Types.getRawType(type)
        if (annotations.isNotEmpty()) return null
        return if (rawType == RealmList::class.java) {
            newRealmListAdapter<RealmObject>(type, moshi)?.nullSafe()
        } else null
    }

    companion object {
        @JvmStatic
        private fun <T : RealmObject?> newRealmListAdapter(type: Type, moshi: Moshi): JsonAdapter<RealmList<T>>? {
            val elementType = Types.collectionElementType(type, RealmList::class.java)
            val elementAdapter: JsonAdapter<T> = moshi.adapter(elementType)
            return RealmListMoshiConverter(elementAdapter)
        }
    }
}